### DISCUSSION TITLE
3D mode cropping buttons

### DISCUSSION THREAD

**Discussion by robinkoehler** [awawa-dev/HyperHDR#64] [Ideas]
Thank you for this excellent fork and for the tutorials! I was wondering if there might be a chance to bring back the 3D mode cropping buttons (none, Side By Side, Top And Bottom)? I understand you had to repurpose that menu for the HDR tonemapping controls. And that I might be the only person in the world actually using the 3D buttons :-)  But it would be neat if it's easy to do. Much appreciated and thanks again!

**Comment by awawa-dev** [comment]
Hi
In fact these modes: (none, Side By Side, Top And Bottom) were separated functionalities that duplicated cropping functionality, even on the video grabber level, because for 'Side By Side' you can use cropping left or right = (video frame width)/2 and for 'Top And Bottom': cropping top or bottom = (video frame height)/2. That was the main reason why there were removed and the buttons were removed alongside them. I would rather go with implementing cropping handling in JSON API and make available to set it remotely: via modern home automation system or just with new api's GET functionality from v16 where they could be even saved as simple bookmarks in the web-browser to set each state.

**Reply by robinkoehler** [reply]
Hi awawa-dev,

Thank you for the prompt response. Ok great, I had no idea those options were available. Still a noob, so pardon if that was obvious.

I would love to create a bookmark to switch from normal cropping to crop bottom 240px. Would you be so kind to point me in the right direction for documentation? Or even provide an example I can modify if you are feeling generous?

Appreciate all your work and the support!

Best wishes,
Robin

**Reply by awawa-dev** [reply]
Hi
'Cropping' is not available right now (and never was) in the remote API, but I think it's not a problem to add it. As I'm now in the middle of important code refactoring I'll add it right after I've done. It will be available to download in from the test build in the 'action' tab on github so you won't need to wait till the next official release. I'll notify you in this topic when it's ready.
Greetings
Awawa

**Reply by robinkoehler** [reply]
Hi Awawa, 

Thanks again for being so responsive. That makes sense! There is no rush, the functionality is available by editing the cropping parameters manually. It would however be a really neat quality of life improvement if and when you get around to it. If it isn't much extra work, in addition to the cropping parameters, the brightness/contrast/saturation/hue parameters would also be a welcome addition. Some movies with elevated black levels seem to trigger the LEDs even though I have set aggressive threshold values. The easy fix is to just lower the brightness of the grabber by a point or two, and having that available via API would be really cool as well. 

Much appreciated! This is an excellent program. It helped me fix the green tint, the red bleed in dark greys, and colour accuracy in a way that I just couldn't with Hyperion NG's functionality. 

Best wishes, 

Robin

**Comment by awawa-dev** [comment]
Hi
Yes, brightness/contrast/saturation/hue parameters from the grabber is next on the list (but it won't be available soon on macOS system due to some current limitations). Personally I want to expose as many parameters & functions as it's possible to external API for ex. for home automation systems. And for brightness/contrast/saturation/hue it's possible to change them even now :) but not on the hardware grabber settings but in the "image processing" tab. It should work with remote commands:
![obraz](https://user-images.githubusercontent.com/69086569/119354308-397bb880-bca4-11eb-93c8-047f7cdf8a9d.png)

All you need to do it's to enable "HyperHDR classic calibration".
They are available at the HyperHDR "remote control" page to change them on fly, but that mode must be enabled in the image processing first to work.

For http GET access: 
1) Prepare command you want:
````
{
  "command":"adjustment",
  "adjustment":{
    "saturationGain":1.1,
    "luminanceGain": 1.2,
    "gammaRed":1.3,
    "gammaGreen":1.4,
    "gammaBlue":1.5
  }
}
````
Encode the result (for ex. at https://www.urlencoder.org/ or any other)

2 run command in the browser and check result on the LED (it won't show in the video preview as it's meant for LED calibration)
````http://HYPERHDRIP:8090/json-rpc?request=```` + result from URL encoding
for example:
http://1HYPERHDRIP:8090/json-rpc?request=%7B%0A%20%20%22command%22%3A%22adjustment%22%2C%0A%20%20%22adjustment%22%3A%7B%0A%20%20%20%20%22saturationGain%22%3A1.1%2C%0A%20%20%20%20%22luminanceGain%22%3A%201.2%2C%0A%20%20%20%20%22gammaRed%22%3A1.3%2C%0A%20%20%20%20%22gammaGreen%22%3A1.4%2C%0A%20%20%20%20%22gammaBlue%22%3A1.5%0A%20%20%7D%0A%7D

Greetings
Awawa

**Reply by robinkoehler** [reply]
HI Awawa, 

Ok great! Appreciate the instructions. I'll have a go later this week. Should be able to make it happen. Thanks for providing the example and for explaining saturation and luminance as well, I wasn't clear how they differed from the grabber settings. 

Best wishes, 

Robin

**Reply by awawa-dev** [reply]
Hi
Cropping method in API is now available in the test release:
https://i.postimg.cc/ZqFpF0Qh/crop.jpg
As you can see the wizard can build the query itself  so no need to do it manually anymore (works also for other commands like controlling brightness)
https://github.com/awawa-dev/HyperHDR/releases/tag/v17.0.0.0beta

**Reply by robinkoehler** [reply]
Hi Awawa, 

Wow, that's great! Amazing!!! Ok, I'll tinker with that on the weekend and will report back. Thank you so much for updating me here in this thread and for adding the new functionality :-)

Take care and best wishes, 

Robin

**Reply by robinkoehler** [reply]
Hi again Awawa,

Absolutely incredible! Installed today (I was SO worried that it would delete all my settings, but everything was backed up and of course it worked without a problem). I had to install it manually via command line, the "install" button inside the HyperHDR interface on the upgrade page was greyed out.

Is there any way I can "chain" commands together, say adjust HDR and cropping? I've tried linking them with an "&" but it failed. I don't know the JSON syntax and if it even accepts two commands in one URL. I have tried

http://192.168.2.3:8090/json-rpc?request=%7B%22command%22:%22videomodehdr%22,%22HDR%22:0%7D&request=%7B%22command%22:%22video-crop%22,%22crop%22:%7B%22left%22:0,%22right%22:3,%22top%22:0,%22bottom%22:2%7D%7D

and that didn't work, so I was obviously doing something wrong. 

But again, this is SO user friendly and absolutely incredible what you're doing there. Sent a few coffees your way as a thankyou.

All the best, 

Robin

**Reply by awawa-dev** [reply]
Hi
Thank you for your support :)
The configuration backup is mainly for situations when for some reason you need downgrade v17 to lower version...it's still possible without a backup but it's a way more complicated. Or when something goes very wrong with v17... ;-)
I've added multi commands execution in latest build: you can find the installers in the latest https://github.com/awawa-dev/HyperHDR/actions If you have some time, please test it. Your example should work now.
The 'install' button was 'inherited': probably I'll remove because it's quite challenging task (and frankly I don't now how to do it for macOS... and each OS should be supported) and pose a security risk when it's not strictly implemented.
Best regards
Awawa

**Reply by robinkoehler** [reply]
Hi Awawa, 

Ok that's awesome! I have shortcuts set up on my phone for now, makes it super easy. The combined commands would be a nice extra, but really not mission critical at the moment. Had a look at the "actions" and with my skillset could not figure out how to install the updates. I'll leave it for now, super happy with the way things are and excited that this just keeps getting better :-)

Thank you and best wishes, 

Robin

![Screenshot_20210827-115037](https://user-images.githubusercontent.com/43112840/131058939-c0e918d4-a0df-4705-96f6-8ba64a0e9ea0.png)

**Reply by robinkoehler** [reply]
> As you can see the wizard can build the query itself so no need to do it manually anymore (works also for other commands like controlling brightness)
> https://github.com/awawa-dev/HyperHDR/releases/tag/v17.0.0.0beta

Hi again Awawa! Congrats on making v17 official :-)  I'm still on beta1 but really appreciate you being so responsive. 
Regarding brightness: I understand this method is available already. I have just encountered a new use case for 3D playback where I have to brighten the source so that I can darken the display. I have to adjust the grabber brightness directly for a usable setup. 

Since implementing the "cropping" on the grabber level was successful, is there any chance to also expose the brightness, contrast, saturation and hue params? I can obviously achieve increased brightness by manually going into the grabber settings in HyperHDR, but it would add yet another layer of convenience, especially now that get requests can be concatenated. 

Thanks for everything, this is a joy to use!

Best wishes, 

Robin

**Reply by awawa-dev** [reply]
Hi @robinkoehler 
I just want to let you know that video controls are exposed in JSON API with the latest commit.
![obraz](https://user-images.githubusercontent.com/69086569/136225826-81ba646e-bc0a-4e3b-b626-0685ccf6f0a1.png)
If you want to test new feature then probably compiling from the latest sources is the best way (Linux installers from HyperHDR github 'Actions' tab are intended for OS based on latest Bullseye).

Best regards
Awawa

**Reply by robinkoehler** [reply]
Hi Awawa, 

This notification slipped through and I didn't reply straight away. Thank you so much for letting me know! Amazing! This just keeps getting better. Some sources just have vastly different input levels, PS5 being the worst offender with delivering a very, very dark signal. This will help automating the correction of that with one simple JSON shortcut :-)

Thanks for everything you do and for the support!

Best wishes, 

Robin

**Comment by awawa-dev** [comment]
Hi Robin
It's possible to include these methods to JSON API. And it should be rather easy for Windows and Linux at least, because version for macOS doesn't support grabber controls. I let you know when it's ready. Post-processing brightness/luminescence/saturation (last 2 options are for 'HyperHDR classic calibration' only) are available in v17 API and you can control them now. 
Best regards
Awawa
